whois domain lookup sensor

GenericStudent opened the pull request:
## Description:
This is at the "test the water" PR stage in that I have developed the basics of the sensor but wish to get feedback on whether this is a suitable component for HA.

I've used the `pythonwhois` library that comes with a `wtfpl` licence, and have pinned the `SCAN_INTERVAL` at 24 hours as we don't want to break the WHOIS TOS with `high volume` requests. I feel that one request per domain every 24 hours should be ok.

Is this a desirable feature and should I fully flesh it out with tests / doc PR?

### Current State
Creates a sensor for each host with the following information:
 - sensor state is the number of days until expiration / Expired if <= 0
 - attributes:
  - name_servers
  - registrar
  - updated (date time)
  - expires (date time)

Any feedback much welcomed :)

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3744

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensors:
  - platform: whois
    domain: example.net
    name: primary
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

GenericStudent committed:
Init commit of new whois sensor

GenericStudent committed:
Updated requirements

GenericStudent committed:
Resolved updated showing expired, added expired attr

GenericStudent committed:
Added missing attribute in init, added whois to coveragerc

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+    def unit_of_measurement(self):
+        """The unit of measurement to present the value in."""
+        if self._expired:
+            return None
```
This will mess-up the history.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+        self.hass = hass
+        self._hostname = hostname
+        self.whois = get_whois
+        self._state = STATE_UNKNOWN
```
Set it to `None` and let Home Assistant handle the rest.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+        """Get the more info attributes."""
+        if self._data:
+            updated_formatted = self._updated_date.strftime(
+                '%Y-%m-%d %H:%M:%S')
```
Please use the ISO 8601 format.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+        if self._expired:
+            return None
+
+        return 'day{}'.format('' if self._state == 1 else 's')
```
The unit of measurement will change if only one day is left. This will make it hard for people who are exporting the data as the sensor will go for `days` to `day` and then to ` `.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+# every 24 hours shouldn't count as "high volume"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): vol.All(cv.ensure_list, [cv.string]),
```
Please allow the user to choose a name for the sensor. Hint: `CONF_NAME`.

GenericStudent replied:
Which style should I go for, Instantiate this sensor with only a single host, such as:
```yaml
sensor:
  - platform: whois
    hostname: example.net
    name: primary
  - platform: whois
    hostname: example.com
    name: secondary
```

or keep the list of hostnames and add the per host name config, such as:
```yaml
sensor:
  - platform: whois
    hostnames:
      - url: example.net
        name: primary
      - url: example.com
        name: secondary
```

I can see merits in both but was curious if there was a process / design used to decide which one to go for.

Thanks for your feedback on this all, much appreciated.

fabaff replied:
I personally prefer the first style (easier for copy-&-paste) but this is pretty much up to the person who is writing the platform.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+        """Initialize the sensor."""
+        from pythonwhois import get_whois
+
+        self.hass = hass
```
Not used.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+
+    def update(self):
+        """Get the current WHOIS data for hostname."""
+        response = self.whois(self._hostname, normalized=True)
```
I guess that there will be an exception if the domain doesn't exist. You need to catch it.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,126 @@
+    devices = []
+
+    for hostname in hostnames:
+        devices.append(WhoisSensor(hass, hostname))
```
You should make sure that the domain exists before setting up the platform otherwise the users will end up if a non-working sensor.

GenericStudent committed:
Various PR comment changes

- Now more resiliant to invalid hostnames
- Removed various assumed STATE_UNKOWN setting
- Upfront check for valid hostname preventing the sensor starting with dud
- Resolved unit of measurement Day, Days, None issue
- Datetime formatting now done to iso 8601 standard
- Removed all expired usage, not really that useful
- Unused hass assignment

GenericStudent committed:
More PR comment resolutions

- Resolved the dilemma with hosts / single host per sensor. Now running
single domain per sensor.
- Renamed host(s) to domain

fabaff commented on .coveragerc:
```diff
@@ -223,6 +223,8 @@ omit =
     homeassistant/components/wemo.py
     homeassistant/components/*/wemo.py
 
+    homeassistant/components/sensor/whois.py
```
Please move it down to the platforms.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,135 @@
+    name = config.get(CONF_NAME)
+
+    try:
+        if 'expiration_date' in get_whois(domain, normalized=True):
```
Especially for European domains is the `expiration_date` not always present. So the lookup is working but we will not get the data.

fabaff replied:
Not much we can do here. Perhaps re-phrasing the warning.

fabaff commented on homeassistant/components/sensor/whois.py:
```diff
@@ -0,0 +1,135 @@
+                self._name_servers = self._data['nameservers']
+
+            self._expiration_date = self._data['expiration_date'][0]
+            self._updated_date = self._data['updated_date'][0]
```
With `home-assistant.io` I get:

```bash
  File "/home/fab/Documents/repos/home-assistant/homeassistant/components/sensor/whois.py", line 131, in update
    self._updated_date = self._data['updated_date'][0]
KeyError: 'updated_date'
```

GenericStudent replied:
Seems my assumption of its existence was incorrect, I've made it conditional now. Resolved the other comments too.

GenericStudent committed:
Moved coveragerc sensor location

GenericStudent committed:
Re-phrased the expiration_date warning

GenericStudent committed:
Resolved assumed updated_date existence

GenericStudent committed:
Resolved missing indent

GenericStudent committed:
Resolved discover_info typo

pvizeli committed:
Update whois.py
